src/solver.rs

```diff
@@ -630,7 +664,276 @@ mod tests {
         }
     }
 
-    // TODO: add tests for ADD
+    const ADD: Instruction = Instruction::Add(RType(0));
+
+    #[test]
+    fn check_invertability_condition_for_add() {
+        let side = ArgumentSide::Lhs;
+
+        test_invertability(ADD, "0", 0b0, 0b0, side, true, "trivial addition");
+        test_invertability(ADD, "0", 0b1, 0b1, side, true, "trivial addition");
+        test_invertability(ADD, "1", 0b0, 0b1, side, true, "trivial addition");
+        test_invertability(ADD, "1", 0b1, 0b10, side, true, "trivial addition");
+        test_invertability(ADD, "1", 0b10, 0b11, side, true, "trivial addition");
+        test_invertability(
+            ADD,
+            "0*",
+            0b10,
+            0b10,
+            side,
+            true,
+            "operand with trailing undetermined bits",
+        );
+        test_invertability(
+            ADD,
+            "0*",
+            0b10,
+            0b11,
+            side,
+            true,
+            "operand with trailing undetermined bits",
+        );
+        test_invertability(
+            ADD,
+            "*0",
+            0b10,
+            0b100,
+            side,
+            true,
+            "operand with leading undetermined bits",
+        );
+        test_invertability(
+            ADD,
+            "*0",
+            0b10,
+            0b11,
+            side,
+            false,
+            "operand with leading undetermined bits",
+        );
+        test_invertability(ADD, "***", 0b100, 0b111, side, true, "undetermined operand");
+        test_invertability(ADD, "***", 0b100, 0b11, side, false, "undetermined operand");
+        test_invertability(
+            ADD,
+            "**0",
+            0b10,
+            0b100,
+            side,
+            true,
+            "operand with multiple undetermined bits",
+        );
+        test_invertability(
+            ADD,
+            "**0",
+            0b1,
+            0b11,
+            side,
+            true,
+            "operand with multiple undetermined bits",
+        );
+        test_invertability(
+            ADD,
+            "**1",
+            0b1,
+            0b111,
+            side,
+            false,
+            "operand with multiple undetermined bits",
+        );
+    }
+
+    #[test]
+    fn compute_inverse_values_for_add() {
+        let side = ArgumentSide::Lhs;
+
+        fn f(l: BitVector, r: BitVector) -> BitVector {
+            l + r
+        }
+
+        // test only for values which are actually invertable
+        test_inverse_value_computation(ADD, "1", 0b1, 0b10, side, f);
+        test_inverse_value_computation(ADD, "0*", 0b10, 0b10, side, f);
+        test_inverse_value_computation(ADD, "***", 0b100, 0b100, side, f);
+        test_inverse_value_computation(ADD, "**0", 0b10, 0b1100, side, f);
+    }
+
+    #[test]
+    fn check_consistency_condition_for_add() {
+        let condition = "t != 0 => x^hi != 0";
```

Comment:
What are these conditions? That's probably a copy-paste error, since these are conditions to check MUL consistency conditions